### PR TITLE
[WIP] VM re-config within existing quota

### PR DESCRIPTION
**Pull request by supernoodz**
Enable quota validation for VM reconfiguration requests. Currently, only new requests are checked.

### CODE REVIEW

**Comment by mkanoor** [comment]
@supernoodz 
Can you please add specs for these changes https://github.com/ManageIQ/manageiq-content/blob/master/spec/automation/unit/method_validation/validate_request_spec.rb

**Comment by tinaafitz** [comment]
@mkanoor We have a WIP PR with the spec test changes:
https://github.com/ManageIQ/manageiq-content/pull/56

File: content/automate/ManageIQ/System/CommonMethods/QuotaMethods.class/__methods__/requested.rb [hunk 1]
```diff
@@ -86,10 +86,16 @@ def vm_prov_option_value(prov_option, options_array = [])
                :flavor        => flavor_obj(@miq_request.get_option(:instance_type)),
                :number_of_vms => get_option_value(@miq_request, :number_of_vms),
                :cloud         => vm_provision_cloud?}
+  # number_of_vms doesn't exist for VmReconfigureRequest
+  args_hash[:number_of_vms] = 1 if @reconfigure_request
```
Comment (mkanoor) [inline content/automate/ManageIQ/System/CommonMethods/QuotaMethods.class/__methods__/requested.rb]: @tinaafitz 
Does this mean a reconfigure can only happen on a single vm and not a group of vms?

File: content/automate/ManageIQ/System/CommonMethods/QuotaMethods.class/__methods__/requested.rb [hunk 2]
```diff
@@ -86,10 +86,16 @@ def vm_prov_option_value(prov_option, options_array = [])
...
 
   case prov_option
   when :vm_memory
-    requested_memory(args_hash, @miq_request.source.vendor)
+    if @reconfigure_request
```
Comment (mkanoor) [inline content/automate/ManageIQ/System/CommonMethods/QuotaMethods.class/__methods__/requested.rb]: @billfitzgerald0120 @tinaafitz 
Can we add a vendor method
```
def vendor
  @reconfigure_request ? @vm.vendor : @miq_request.source.vendor
end
```

The this becomes
```
   when :vm_memory
        requested_memory(args_hash, vendor)
```

File: content/automate/ManageIQ/System/CommonMethods/QuotaMethods.class/__methods__/requested.rb [hunk 3]
```diff
@@ -112,6 +127,17 @@ def requested_number_of_cpus(args_hash)
...
+
+  if @reconfigure_request && args_hash[:resource].options[:number_of_sockets]
+    # Account for the VM's existing CPUs
+    args_hash[:prov_value] = args_hash[:prov_value].to_i - @vm.hardware.cpu_total_cores.to_i \
+      * @vm.hardware.cpu_cores_per_socket.to_i
+
+    $evm.log(:info, "vm_number_of_cpus:         #{@vm.hardware.cpu_total_cores.to_i \
+      * @vm.hardware.cpu_cores_per_socket.to_i}")
+    $evm.log(:info, "requested_number_of_cpus:  #{args_hash[:prov_value].to_i}")
+    @bypass_quota = false if args_hash[:prov_value].to_i > 0
```
Comment (mkanoor) [inline content/automate/ManageIQ/System/CommonMethods/QuotaMethods.class/__methods__/requested.rb]: @billfitzgerald0120 @tinaafitz 
The bypass_quota seems like an incorrect variable name here. Should it rather be
```
 @check_quota = true if args_hash[:prov_value].to_i > 0
```

If its a reconfigure request we can set the
```
@check_quota = false
```
And have one of these methods enable it, if the memory, cpu, storage in reconfigure request is greater than the currently allocated amount.

File: content/automate/ManageIQ/System/CommonMethods/QuotaMethods.class/__methods__/requested.rb [hunk 4]
```diff
@@ -120,8 +146,35 @@ def vmdb_object(model, id)
...
+    if args_hash[:resource].options[:disk_add]
+      args_hash[:resource].options[:disk_add].each do |disk|
+        $evm.log(:info, "Adding a disk:  #{disk.inspect}")
+        args_hash[:prov_value] += disk['disk_size_in_mb'].to_i.megabytes
+      end
+    end
+    if args_hash[:resource].options[:disk_remove]
+      args_hash[:resource].options[:disk_remove].each do |disk|
+        $evm.log(:info, "Removing disk: #{disk.inspect}")
+        if disk_num == disk[:disk_name].match(/_(\d).vmdk/)
```
Comment (mkanoor) [inline content/automate/ManageIQ/System/CommonMethods/QuotaMethods.class/__methods__/requested.rb]: @supernoodz @billfitzgerald0120 
How is disk_num being set? We are trying to compare a value which is not set

**Comment by mkanoor** [comment]
Integrated into https://github.com/ManageIQ/manageiq-content/pull/56
Hence closing this PR

**Pull request closed without merging**